recipe/recipe.go

```diff
@@ -36,3 +42,335 @@ type instance struct {
+			if len(instanceInvalidFields) > 0 {
+				for _, iField := range instanceInvalidFields {
+					instanceInvalidFields[i] = "output-instances[" + strconv.Itoa(i) + "]." + iField
+				}
+			}
+			missingFields = append(missingFields, instanceMissingFields...)
+			invalidFields = append(invalidFields, instanceInvalidFields...)
+		}
+	} else {
+		missingFields = append(missingFields, "output-instances")
+	}
+
+	if missingFields != nil {
+		return fmt.Errorf("missing mandatory fields: %v", missingFields)
+	}
+
+	if invalidFields != nil {
+		return fmt.Errorf("invalid fields: %v", invalidFields)
+	}
+
+	return nil
+
+}
+
+//ValidateAddInstance - checks if fields of OutputInstances are not empty
+func (instance *Instance) ValidateAddInstance() error {
+	var missingFields []string
+	var invalidFields []string
+
+	instanceMissingField, instanceInvalidField := instance.validateInstance()
+	missingFields = append(missingFields, instanceMissingField...)
+	invalidFields = append(invalidFields, instanceInvalidField...)
+
+	if missingFields != nil {
+		return fmt.Errorf("missing mandatory fields: %v", missingFields)
+	}
+
+	if invalidFields != nil {
+		return fmt.Errorf("invalid fields: %v", invalidFields)
+	}
+
+	return nil
+}
+
+//ValidateAddCodelist - checks if fields of Codelist are not empty
+func (codelist *CodeList) ValidateAddCodelist() error {
+	var missingFields []string
+	var invalidFields []string
+
+	codelistMissingFields, codelistinvalidFields := codelist.validateCodelists()
+	missingFields = append(missingFields, codelistMissingFields...)
+	invalidFields = append(invalidFields, codelistinvalidFields...)
+
+	if missingFields != nil {
+		return fmt.Errorf("missing mandatory fields: %v", missingFields)
+	}
+
+	if invalidFields != nil {
+		return fmt.Errorf("invalid fields: %v", invalidFields)
+	}
+
+	return nil
+}
+
+//ValidateUpdateRecipe - checks updates of recipe for PUT request
+func (recipe *Response) ValidateUpdateRecipe() error {
+	var missingFields []string
+	var invalidFields []string
+	validFormats := map[string]bool{
+		"v4": true,
+	}
+
+	// Validation to check at least one field of the recipe is updated
+	allStringFieldsEmpty := recipe.ID == "" && recipe.Format == "" && recipe.Alias == ""
+	allRemainingFieldsNil := recipe.InputFiles == nil && recipe.OutputInstances == nil
+	if allStringFieldsEmpty && allRemainingFieldsNil {
+		invalidFields = append(invalidFields, "no recipe fields updates given")
+	}
+
+	if recipe.ID != "" {
+		invalidFields = append(invalidFields, "id cannot be changed")
+	}
+	if recipe.Format != "" {
+		if !validFormats[recipe.Format] {
+			invalidFields = append(invalidFields, "format is not valid")
+		}
+	}
+
+	if recipe.InputFiles != nil && len(recipe.InputFiles) > 0 {
+		for i, file := range recipe.InputFiles {
+			if file.Description == "" {
+				missingFields = append(missingFields, "input-files["+strconv.Itoa(i)+"].description")
+			}
+		}
+	}
+	if recipe.InputFiles != nil && len(recipe.InputFiles) == 0 {
```

Comment:
This might add complexity, but we could actually make updating the input files and output instances optional? if they provide any of those details, then we can just replace the existing full list with the new valid full list, but if they dont provide it we maybe don't have to fail? We'd need to update the UpdateRecipe handler to make sure the update doesnt wipe those fields I guess

Otherwise, I would just expect a very clear comment in the swagger spec that when updating a recipe, a complete list of input files and output instances must be provided

Reply:
This is how it is implemented. When updating the recipe, updating the input files and output instances within that recipe are optional. 

When a user does _not_ give information to update input files or output instances, then `recipe.InputFiles` or `recipe.OutputInstances` are `nil`. As their types are both set to `omitempty`, their fields won't be overwritten when `nil`.

On the other hand, if a user does give updates to input files or output instances, then the entire `recipe.InputFiles` or  `recipe.OutputInstances` would be overwritten to the value given by user in which we just replace the existing full list with the new full list, as you mentioned. But at this point, it is important to validate that the user has given _valid_ values to _all_ fields of the `recipe.InputFiles` or  `recipe.OutputInstances` and this is carried out by the _ValidateUpdateRecipe(ctx)_

So for example; in the if statement "`if recipe.InputFiles != nil && len(recipe.InputFiles) == 0`", it gives an invalid field error as the user has given an update to recipe.InputFiles (as it is `not nil`) but the update that he has given is an empty array (so `len(recipe.InputFiles)==0`). But it won't fail if `recipe.InputFiles `is `nil` because no form of validation is given in this case as it is automatically ignored by the `omitempty`